src/sentinel.c

```diff
@@ -3837,11 +3840,31 @@ void sentinelCommand(client *c) {
             addReplyBulkLongLong(c, addr->port);
         }
     } else if (!strcasecmp(c->argv[1]->ptr, "failover")) {
-        /* SENTINEL FAILOVER <primary-name> */
+        /* SENTINEL FAILOVER <primary-name> [type <legacy | pause>] */
```

Comment:
Do we really need to let the user select the legacy behaviour? Who wants data loss?

Is it a breaking change if we just change this to always use FAILOVER? (a fallback to the old behavior if FAILOVER is not supported is OK)

Reply:
I am worried that the primary support FAILOVER, but the replica does not. In this case, the primary seems to return OK, and sentinel has no way of knowing whether the replica support FAILOVER. And then the failover timeout and always timeout in every retry. Not sure about this case. Or we can default to new way and provide option to use the old way if user find out the failover is always timeout?

If we provide options, maybe we should not have this fallback logic seems it may cause trouble if the user really want to use the new way and find out it is using old way at the end. (As long as it is well documented, i guess it should be fine.)

Reply:
In valkey 8, the replica sends `REPLCONF version`. If it's valkey 8 or later we can use failover. If it's older we can use replica of noone?

Reply:
The REPLCONF version is only sent to primary from a replica, right? You are saying that in FAILOVER, we check the replica version? This require a change in primary side, like we need to do some coding around it and make FAILOVER reject with an old replica.

What happend if a new sentinel send a FAILOVER TO to a primary (without the replica version check)? And this will also break FAILOVER command in some area, i don't think we should limit it.

When FAILOVER is added, we extend PSYNC to take a FAILOVER option in the replica, but we don't extend REPLCONF capa to support it. So a primary don't have a way to know if replica support PSYNC FAILOVER. In this case, primary will abort the FAILOVER.

Added these options somehow i pass the burden of choice to the user (to avoid trouble on myself..), who can choose the right option based on there version, and can use the old way when the new way fails.

Reply:
Ahhh, I mixed up with the shutdown-failover feature. :laughing:

I prefer to avoid a new config if we can. Sorry for the trouble.

If the primary supports FAILOVER but the replica doesn't support FAILOVER, it means the primary is newer than the replica. We don't support this configuration, I think. In a rolling upgrade, the new version should always be added as a replica first. At least that's how to rolling upgrade in a cluster.

The FAILOVER command was added in 6.2. Even if we want to support downgrade, IMO we don't need to support downgrade to < 6.2, because 6.0 is already end-of-life.

So, I think it's acceptable that sentinel tries FAILOVER and if the primary doesn't support it, it fallbacks to REPLICAOF NO ONE. You don't think so?

Another alternative I guess is that sentinel can coordinate the failover:

1. Sentinel sends CLIENT PAUSE WRITE to the primary
2. Sentinel checks offset of the replica
3. Sentinel sends REPLICAOF

Reply:
> So, I think it's acceptable that sentinel tries FAILOVER and if the primary doesn't support it, it fallbacks to REPLICAOF NO ONE. You don't think so?

i think it is ok, i am ok with the fallback logic. So i just need to keep this PR and drop the legacy type, right? You are right, i thought too much about this in this version thing..

Yes, the client pause is another option, it require more code, i think FAILOVER is somehow easy to use so i just use it, it also has some INFO fields we can see unless CLIENT PAUSE. The benefit of Client PASUE may be that it can work on old Redis (maybe 3.0?). Which one do you think is better?

Reply:
Less work is better. :)

> So i just need to keep this PR and drop the legacy type, right? 

Yes.